src/include/migraphx/op/pooling.hpp

```diff
@@ -57,59 +60,116 @@ struct pooling
+                    if(padding_size == 2 * kdims)
+                        padding_factor = padding[i] + padding[i + kdims];
+                    assert(input_lens[i + 2] + padding_factor >= lengths[i]);
+                    std::size_t dim_size = input_lens[i + 2] + padding_factor - lengths[i];
+                    std::size_t len      = (ceil_mode)
+                                          ? dim_size / stride[i] +
+                                                static_cast<std::size_t>(
+                                                    (dim_size % stride[i] != 0)) // ceil uint divide
+                                          : dim_size / stride[i];                // floor divide
+                    output_lens.push_back(len + 1);
+                }
+            }
+            return output_lens;
+        };
 
-        for(size_t i = 0; i < kdims; i++)
+        if(input.dynamic())
         {
-            std::ptrdiff_t dim_size;
-            auto padding_factor = 2 * padding[i];
-            if(padding_size == 2 * kdims)
-                padding_factor = padding[i] + padding[i + kdims];
-            dim_size = input_lens[i + 2] + padding_factor - lengths[i];
-            assert(dim_size >= 0);
-            std::size_t len = (ceil_mode) ? ceil_divide<std::ptrdiff_t>(dim_size, stride[i])
-                                          : floor_divide<std::ptrdiff_t>(dim_size, stride[i]);
-
-            output_lens.push_back(std::size_t(std::max<std::ptrdiff_t>(1, len + 1)));
+            auto input_dyn_dims = input.dyn_dims();
+            std::vector<shape::dynamic_dimension> output_dyn_dims(input_dyn_dims.begin(),
+                                                                  input_dyn_dims.begin() + 2);
+            if(dyn_global)
+            {
+                for(size_t i = 0; i < kdims; ++i)
+                {
+                    output_dyn_dims.push_back(shape::dynamic_dimension{1, 1, 1});
+                }
+                return {input.type(), output_dyn_dims};
+            }
+            else
+            {
+                auto min_spatial_dims = calc_spatial_dim_out(input.min_lens());
+                auto max_spatial_dims = calc_spatial_dim_out(input.max_lens());
+                auto opt_spatial_dims = calc_spatial_dim_out(input.opt_lens());
+                for(size_t i = 0; i < kdims; ++i)
+                {
+                    output_dyn_dims.push_back(shape::dynamic_dimension{
+                        min_spatial_dims[i], max_spatial_dims[i], opt_spatial_dims[i]});
+                }
+                return {input.type(), output_dyn_dims};
+            }
         }
-        return inputs[0].with_lens(output_lens);
-    }
+        else
+        {
+            auto input_lens = input.lens();
 
-    size_t kdims() const
-    {
-        check_attribute_size();
-        return stride.size();
+            std::vector<std::size_t> output_lens(input_lens.begin(), input_lens.begin() + 2);
+            if(dyn_global)
```

Comment:
Why is this checked when the input is not dynamic? If its not dynamic then we should know if the pooling is global based on window size.

Reply:
Mistake, shouldn't be checked here.

Reply:
I remember why it's here now. It's for when the `compute_shape()` is called again at model evaluation time by the `dynamic_output` object. The kernel size is not used for that case.